src/gui/dialogs/DialogStateful.cpp

```diff
@@ -0,0 +1,131 @@
+#include "DialogStateful.hpp"
+#include "DialogRadioButton.hpp"
+#include "gui.h"
+
+extern window_t *window_1; //current popup window, C-code remain
+
+//*****************************************************************************
+
+IDialogStateful::IDialogStateful(const char *name, int16_t WINDOW_CLS_)
+    : IDialog(WINDOW_CLS_)
+    , id_capture(window_capture())
+    , color_back(gui_defaults.color_back)
+    , color_text(gui_defaults.color_text)
+    , font(gui_defaults.font)
+    , font_title(gui_defaults.font_big)
+    , padding(gui_defaults.padding)
+    , flags(0)
+    , last_text_h(0)
+    , phase(0)
+    , progress(-1)
+    , title(name) {
+    window_1 = this; //todo
+    gui_reset_jogwheel();
+    gui_invalidate();
+    window_set_capture(id);
+}
+
+bool IDialogStateful::Change(uint8_t phs, uint8_t progress_tot, uint8_t progr) {
+    if (!can_change(phs))
+        return false;
+    if (phase != phs) {
+        phase = phs;
+        flags |= DLG_PHA_CH;
+    }
+    if (progress_tot != progress) {
+        progress = progress_tot;
+        flags |= DLG_PRO_CH;
+    }
+    gui_invalidate();
+    return true;
+}
+
+IDialogStateful::~IDialogStateful() {
+    window_destroy(id);
+    window_set_capture(id_capture);
+    window_invalidate(0);
+}
+
+extern window_t *window_1; //current popup window
+
+void IDialogStateful::draw_frame() {
+    rect_ui16_t rc = rect;
+    display->draw_line(point_ui16(rc.x, rc.y), point_ui16(239, rc.y), COLOR_GRAY);
+    display->draw_line(point_ui16(rc.x, rc.y), point_ui16(rc.x, 320 - 67), COLOR_GRAY);
+    display->draw_line(point_ui16(239, rc.y), point_ui16(239, 320 - 67), COLOR_GRAY);
+    display->draw_line(point_ui16(rc.x, 320 - 67), point_ui16(239, 320 - 67), COLOR_GRAY);
+}
+
+//this should be moved elswhere
+void progress_draw(rect_ui16_t win_rect, font_t *font, color_t color_back,
+    color_t color_text, padding_ui8_t padding, uint8_t progress) {
+    rect_ui16_t rc_pro = win_rect; //must copy it
+    char text[16];
+    rc_pro.x += 10;
+    rc_pro.w -= 20;
+    rc_pro.h = 16;
+    rc_pro.y += 30;
+    uint16_t w = rc_pro.w;
+    rc_pro.w = w * progress / 100;
+    display->fill_rect(rc_pro, COLOR_ORANGE);
+    rc_pro.x += rc_pro.w;
+    rc_pro.w = w - rc_pro.w;
+    display->fill_rect(rc_pro, COLOR_GRAY);
+    rc_pro.y += rc_pro.h;
+    rc_pro.w = win_rect.w - 120;
+    rc_pro.x = win_rect.x + 60;
+    rc_pro.h = 30;
+    sprintf(text, "%d%%", progress);
+    render_text_align(rc_pro, text, font, color_back, color_text, padding, ALIGN_CENTER);
+}
+//this should be moved elswhere
```

Comment:
Use TODO in such comment, it can be find easily.

Reply:
fixed